fix(supabase): get jwt on http call

Vinzent03 opened the pull request:
## What is the new behavior?

I'm now getting the latest access token when the actual request is done by wrapping the http client with another client, which gets the newest from `gotrue` and sets the headers correctly.

Additionally I made `expiresAt` a late variable to avoid reparsing the jwt each time. And I added `expired`, because I think its a nice addition to easy check the validity of a session. But I'm not sure if some additional padding could be useful. So that a session is maybe 5 seconds before actual expiration, marked as expired.

## Additional context
That technique is inspired from supabase-js

dshukertjr reviewed (commented):
This is brilliant!! I attempted to do what the js client library was doing, but I couldn't find a way to do it without breaking change, but here you are to save the day again! I am loving every bit of the code in this PR! Thank you so much for the amazing work!

Vinzent03 commented on packages/gotrue/lib/src/types/session.dart:
```diff
@@ -66,6 +68,12 @@ class Session {
     }
   }
 
+  bool get expired {
```
Great!

Vinzent03 commented:
Thanks for the kind words 😊 
Updated to `isExpired` and added the buffer.
One thing though that came to my mind are race conditions. If the session is expired and one does `Future.wait(<listOfFutures>)` all, for example postgrest, calls are done "simultaneously". I think this may make the session refresh for each request. Should there be some catching for this or am I missing something?

dshukertjr commented:
>Thanks for the kind words 😊
Updated to isExpired and added the buffer.
One thing though that came to my mind are race conditions. If the session is expired and one does Future.wait(<listOfFutures>) all, for example postgrest, calls are done "simultaneously". I think this may make the session refresh for each request. Should there be some catching for this or am I missing something?

Yeah this PR is a bit scary to publish without a more robust integration tests, ins't it? I know that gotrue server has a mechanism to allow token refresh using the same access token for a short period of time (I believe it was 10 seconds) to allow simultaneous token refreshes to happen. 

Either way, I was thinking about creating some additional tests within the supabase_flutter package that launches a local Supabase instance using the [Supabase Github actions](https://github.com/supabase/supabase-action-example/blob/main/.github/workflows/ci.yaml) and tests basic use cases of all the Supabase components like postgrest or storage from a Flutter app. Then we could include a test case where we try to sending bunch of requests using `Future.await` and see if it works properly. What do you think about this?

dshukertjr commented:
>The only thing that comes to my mind that changes in flutter environment vs pure dart is the issue of an app going to background or network issues, but these are hard to test.

Probably testing realtime would be easier too, I think. But yeah maybe not a huge gain overall as you say.

Vinzent03 commented:
My guess was indeed the case and I've added a global completer now to combine these token refresh requests. When trying to implement that, I removed all the other completer that were used for the retry logic. All tests are passing and I think everything is fine. The changes of this pr are more significant than expected, though. I think the code of the retry logic is cleaner as well.

Vinzent03 commented on packages/gotrue/lib/src/gotrue_client.dart:
```diff
@@ -498,13 +501,11 @@ class GoTrueClient {
 
-    _callRefreshToken(refreshCompleter);
-    return refreshCompleter.future;
+    return await _callRefreshToken();
```
The idea behind adding that await there is to build a better stacktrace. If that future fails and is not awaited there, I think the `refreshSession` method wouldn't be contained in the stacktrace.

dshukertjr replied:
Oh, this is an interesting trick! Awesome!

dshukertjr commented on packages/gotrue/lib/src/gotrue_client.dart:
```diff
@@ -863,30 +863,29 @@ class GoTrueClient {
         accessToken: accessToken,
       );
+      return _refreshTokenCompleter!.future;
     } catch (error, stack) {
       if (error is AuthException) {
         if (error.message == 'Invalid Refresh Token: Refresh Token Not Found') {
           await signOut();
         }
       }
-      completer.completeError(error, stack);
       _onAuthStateChangeController.addError(error, stack);
+      rethrow;
```
I think developers can listen to the errors with `onStateChanged`, so I'm not too concerned, but thought I would flag it with you!

Vinzent03 replied:
I've added a new commit to catch any errors from `_callRefreshToken` in `_saveSesison` and `_setTokenRefreshTimer`, because these should already be added to the stream and be handled there. Testing this with a test is very complicated, because all these methods are private and I would have to create a cusomt http client to mock some responses.

dshukertjr replied:
Thank you so much for all the hard work!

dshukertjr replied:
I wouldn't mind making things public with the [internal](https://api.flutter.dev/flutter/meta/internal-constant.html) annotation like you did with another method or property the other day if that makes testing things easier.

Vinzent03 replied:
I don't think that extra effort is neccessary. I'm very confident the errors are properly catched. `_saveSession` and `_setTokenRefreshTimer` are the only ones calling `_callRefreshToken`, which are not directly called by the user, so I'm pretty sure that's correct now.

dshukertjr commented:
Looks great to me. If you are ready to merge it, let's merge it and we can publish new versions. .
